### PR TITLE
Add shopper Tracks events

### PR DESCRIPTION
**Pull request by malithsen**
Fixes https://github.com/Automattic/woopay/issues/1907

#### Changes proposed in this Pull Request

This PR adds the following Tracks events
```
wcpay_cart_page_view
wcpay_product_page_view
wcpay_gpay_button_click
wcpay_apple_pay_button_click
wcpay_checkout_email_address_woopay_check
wcpay_checkout_woopay_save_my_info_tooltip_hover
wcpay_checkout_woopay_save_my_info_tooltip_learn_more_click
wcpay_checkout_woopay_save_my_info_privacy_policy_click
wcpay_checkout_save_my_info_tos_click
```

#### Testing instructions

Save My Info events

* Add a product to the cart
* Visit shortcode checkout page as an unregistered WooPay user.
* Hover over the "i" tooltip in Save My Info section.
* Click the Learn More link in the tooltip. (`wcpay_checkout_woopay_save_my_info_tooltip_hover`)
* Check the save my info checkbox. 
* Click on Terms of Service and Privacy Policy links.  (`wcpay_checkout_save_my_info_tos_click`, `wcpay_checkout_woopay_save_my_info_privacy_policy_click`)
* In ~5 minutes, check MC Tracks live view and make sure above events are recorded.
* Repeat the tests for Blocks checkout.

Payment Request Buttons

* Click on the Google Pay button from product, cart and checkout page.
* Click on the Apple Pay button from the product, cart and checkout page.
* Check Tracks Live view for `wcpay_gpay_button_click` and `wcpay_apple_pay_button_click` with the relevant `source` property indicating the page.

Page Views
* View the cart page.
* View a product page.
* Check MC Live view for `wcpay_cart_page_view` and `wcpay_product_page_view` events.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7268` or branch name `add/shopper-tracks-events` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=add/shopper-tracks-events) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: f718230aa2297115ddd151c657dc2134fb9bda77
- Build time: 2023-10-09 20:19:15 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: client/payment-request/blocks/payment-request-express.js [hunk 1]
```diff
@@ -53,9 +53,14 @@ const PaymentRequestExpressComponent = ( {
...
 
+	const onPaymentRequestButtonClick = () => {
+		console.log( 'clicked' );
+		onButtonClick();
+	};
+
 	return (
 		<PaymentRequestButtonElement
-			onClick={ onButtonClick }
+			onClick={ onPaymentRequestButtonClick }
```
Comment (alefesouza) [inline client/payment-request/blocks/payment-request-express.js]: This looks like debug code.

**Review by alefesouza — APPROVED** [review]
Code looks good and events are emitted as expected, just a minor change for a code that probably should not be committed.

File: client/payment-request/blocks/payment-request-express.js (hunk 1 above)
Comment (malithsen) [reply client/payment-request/blocks/payment-request-express.js]: 🤦 Thanks for catching that!
I had some uncommitted changes. To test them please follow instructions under "Payment Request Buttons" on the Blocks checkout page.

Comment (malithsen) [reply client/payment-request/blocks/payment-request-express.js]: @alefesouza I am requesting another review due to the above changes :)

File: client/payment-request/blocks/payment-request-express.js [hunk 2]
```diff
@@ -53,9 +54,45 @@ const PaymentRequestExpressComponent = ( {
...
+			paymentRequestType = 'apple_pay';
+		} else if ( result.googlePay ) {
+			paymentRequestType = 'google_pay';
+		}
+	} );
+
+	const onPaymentRequestButtonClick = () => {
+		onButtonClick();
+
+		if ( paymentRequestType === 'google_pay' ) {
```
Comment (eduardoumpierre) [inline client/payment-request/blocks/payment-request-express.js]: **Non-blocking:** What do you think about grouping those payment request types in an object? By doing this we can map the payment request type and its event. We could have something like this:

```js
const paymentRequestTypeEvents = {
  google_pay: wcpayTracks.events.GOOGLEPAY_BUTTON_CLICK,
  apple_pay: wcpayTracks.events.APPLEPAY_BUTTON_CLICK,
};

if ( paymentRequestTypeEvents.hasOwnProperty( paymentRequestType ) ) {
  const event = paymentRequestTypeEvents[ paymentRequestType ];
  wcpayTracks.recordUserEvent( event, { source: 'checkout' } );
}
```

File: client/payment-request/index.js [hunk 3]
```diff
@@ -341,6 +342,21 @@ jQuery( ( $ ) => {
 			const addToCartButton = $( '.single_add_to_cart_button' );
 
 			prButton.on( 'click', ( evt ) => {
+				if ( paymentRequestType === 'google_pay' ) {
```
Comment (eduardoumpierre) [inline client/payment-request/index.js]: We may benefit from using the same structure here too.

File: client/payment-request/blocks/payment-request-express.js (hunk 2 above)
Comment (malithsen) [reply client/payment-request/blocks/payment-request-express.js]: Thanks @eduardoumpierre That's a great suggestion!
Updated in https://github.com/Automattic/woocommerce-payments/pull/7268/commits/f392390c45ba64bcab8dc7ea6d8c3d4c7f45a5d7

There's still a bit of duplicated code between blocks and classic pages, while we can move `trackPaymentRequestButtonClick` to `wcpayTracks` to eliminate that, I think we should limit `wcpayTracks` to high-level Tracks functions. So opting to keep what's in https://github.com/Automattic/woocommerce-payments/pull/7268/commits/f392390c45ba64bcab8dc7ea6d8c3d4c7f45a5d7 as is :)

Comment (eduardoumpierre) [reply client/payment-request/blocks/payment-request-express.js]: Just to confirm: The source content from the second function on `client/payment-request/index.js` changed from `wcpayPaymentRequestParams.button_context` to `checkout`, was it intended?

Comment (malithsen) [reply client/payment-request/blocks/payment-request-express.js]: Good catch. That wasn't intentional. Fixed in https://github.com/Automattic/woocommerce-payments/pull/7268/commits/3277a2dc4b6eef96da4234eead7a2828496df8e0

**Review by alefesouza — APPROVED** [review]
LGTM, it worth mentioning that the event emitted on Google Pay click is `wcpay_googlepay_button_click`, not ´ wcpay_gpay_button_click´ as seen on the PR description.

**Comment by malithsen** [comment]
Thanks @alefesouza fixed the event name in https://github.com/Automattic/woocommerce-payments/pull/7268/commits/0287dc7f52a97f955b4454ff66bb3557d1de96b2

**Pull request merged**